mlflow/pyspark/optuna/study.py

```diff
@@ -0,0 +1,264 @@
+            value_or_values = func(trial)
+        except exceptions.TrialPruned as e:
+            state = TrialState.PRUNED
+            func_err = e
+        except (Exception, KeyboardInterrupt) as e:
+            state = TrialState.FAIL
+            func_err = e
+            func_err_fail_exc_info = traceback.format_exc()
+        try:
+            frozen_trial, warning_message = optuna.study._tell._tell_with_warning(
+                study=study,
+                trial=trial,
+                value_or_values=value_or_values,
+                state=state,
+                suppress_warning=True,
+            )
+        except Exception:
+            frozen_trial = study._storage.get_trial(trial._trial_id)
+            warning_message = None
+
+        if frozen_trial.state == TrialState.COMPLETE:
+            study._log_completed_trial(frozen_trial)
+        elif frozen_trial.state == TrialState.PRUNED:
+            _logger.info("Trial {} pruned. {}".format(frozen_trial._trial_id, str(func_err)))
+            mlflow_client.set_terminated(frozen_trial._trial_id, status="KILLED")
+        elif frozen_trial.state == TrialState.FAIL:
+            error_message = None
+            if func_err is not None:
+                error_message = func_err_fail_exc_info
+            elif warning_message is not None:
+                error_message = warning_message
+            with tempfile.TemporaryDirectory() as tmp_dir:
+                path = Path(tmp_dir, "error_message.txt")
+                path.write_text(error_message)
+                # Log the file as an artifact in the active MLflow run
+                mlflow_client.log_artifact(frozen_trial._trial_id, path)
+                mlflow_client.set_terminated(frozen_trial._trial_id, status="FAILED")
+
+        if (
+                frozen_trial.state == TrialState.FAIL
+                and func_err is not None
+                and not isinstance(func_err, catch)
+        ):
+            raise func_err
+
+        if callbacks is not None:
+            for callback in callbacks:
+                callback(study, frozen_trial)
+
+
+class MlflowSparkStudy(Study):
+    """A wrapper of :class:`~optuna.study.Study` to incorporate Optuna with spark
+    via MLflow experiment.
+
+    .. code-block:: python
+        :caption: Example
+
+        from mlflow.optuna.storage import MlflowStorage
+        from mlflow.pyspark.optuna.study import MlflowSparkStudy
+
+
+        def objective(trial):
+            x = trial.suggest_float("x", -10, 10)
+            return (x - 2) ** 2
+
+
+        experiment_id = "507151065975140"
+        study_name = "spark_mlflow_storage"
+
+        storage = MlflowStorage(experiment_id=experiment_id)
+        mlflow_study = MlflowSparkStudy(study_name, storage)
+        mlflow_study.optimize(objective, n_trials=4)
+    """
+
+    def __init__(
+        self,
+        study_name: str,
+        storage: MlflowStorage,
+        sampler: Optional["samplers.BaseSampler"] = None,
```

Comment:
```suggestion
        sampler: Optional[samplers.BaseSampler] = None,
```

we don't need to use forward ref.